lib/Dialect/TritonGPU/Transforms/CombineTensorSelectAndIf.cpp

```diff
@@ -38,10 +39,51 @@ class CombineTensorSelectAndIfPass
   void runOnOperation() override {
     MLIRContext *context = &getContext();
     ModuleOp m = getOperation();
-    DominanceInfo dom(m);
+
+    // The user of select maybe inside either the ThenRegion or ElseRegion of
+    // the scf.if. So, canonicalize user of select in scf.if first.
+    llvm::MapVector<std::pair<Value, Value>, SmallVector<Operation *>>
+        usersNeedreplaced;
+    m.walk([&](arith::SelectOp selectOp) {
+      auto *parentBlock = selectOp->getBlock();
+      Value condition = selectOp.getOperand(0);
+      Value trueVal = selectOp.getOperand(1);
+      Value falseVal = selectOp.getOperand(2);
+      Value resVal = selectOp.getResult();
+      for (auto *condUser : condition.getUsers()) {
+        auto ifOp = dyn_cast<scf::IfOp>(condUser);
+        // Only consider scf.if within the same block as arith.select.
+        if (!ifOp || ifOp->getBlock() != parentBlock)
+          continue;
+        for (auto *resUser : resVal.getUsers()) {
+          if (ifOp->isProperAncestor(resUser)) {
+            if (ifOp.getThenRegion().findAncestorOpInRegion(*resUser) !=
+                nullptr) {
+              // The user is inside the ThenRegion of the scf.if.
+              usersNeedreplaced[std::make_pair(resVal, trueVal)].push_back(
+                  resUser);
+            } else {
+              // The user is inside the ElseRegion of the scf.if.
+              usersNeedreplaced[std::make_pair(resVal, falseVal)].push_back(
+                  resUser);
+            }
+          }
+        }
+      }
+    });
+    for (auto [replacedSrcAndDst, users] :
+         llvm::make_early_inc_range(usersNeedreplaced)) {
+      Value srcVal = replacedSrcAndDst.first;
+      Value dstVal = replacedSrcAndDst.second;
+      for (Operation *user : llvm::make_early_inc_range(users)) {
+        srcVal.replaceUsesWithIf(
+            dstVal, [&](OpOperand &use) { return use.getOwner() == user; });
+      }
+    }
```

Comment:
You could make this a rewrite pattern instead, it should look much simpler. 
It's interesting MLIR doesn't have this optimization.

Reply:
done~